### PR TITLE
Add a workaround to allow SHA-1 signatures in certs to work for Vault <= 1.11

### PR DESCRIPTION
**Pull request by swenson**
**tl;dr** I'm sorry

This will allow us to upgrade to Go 1.19+ in Vault 1.10 and 1.11, which helps keeps security tools quiet about using older versions with CVE.

## Why
[Go 1.18+ has removed support (by default) for certificates containing signatures with SHA-1.](https://go.dev/doc/go1.18#sha1)
For Vault 1.12+, we've [officially deprecated](https://developer.hashicorp.com/vault/docs/deprecation/faq#q-what-is-the-impact-of-removing-support-for-x-509-certificates-with-signatures-that-use-sha-1) these kinds of signatures and have a runtime workaround using an environment variable that users can set. This is the only official workaround.

## What does this PR do?
This PR patches an internal Go variable (that is also patched by the environment variable mentioned above), but does so permanently and only in code, so that users don't have to set the environment variable.

It only applies the patch is the detected Vault version is <= 1.11.0.

## That's terrible. What other options are there?

There are essentially 4.5 options:

1. Do nothing, don’t upgrade to Go 1.18 or 1.19, and let security scanners be mad while we don’t upgrade Go to 1.19 in Vault 1.10 and 1.11.
2. (This PR) Patch Go 1.19 at runtime to allow SHA-1-based signatures. This has the least impact on users, but is fragile and could stop working at any point since it relies on manipulating internal variables in Go. We would have to maintain this until Vault 1.12 is the oldest supported release in a year or so.
3. Fork `crypto/x509` to continue allowing SHA-1. This is potentially a lot of work, and we’d probably just want to throw away our fork when Vault 1.12 is the oldest supported release.

   a. Alternatively but similarly, we could fork and patch Go itself and use our fork for builds, like how the BoringCrypto fork works. This is probably an even worse idea due to the maintenance burden of maintaining a Go fork for all of the platforms that Vault supports.

4. Upgrade to Go 1.19, but don’t do this PR. This could break users unless they use the environment variable workaround. This probably breaks our [deprecation policy](https://developer.hashicorp.com/vault/docs/deprecation/faq#q-what-are-the-phases-of-deprecation).

If we cannot do option 1, I believe that option 2 (this PR) is the path of least pain and suffering for our users, with hopefully little pain on our part.

## What is the risk of doing this?

If the Go team decides to change the way they implement the SHA-1 deprecation workaround by something as simple as changing the internal variable name they use, we're going to have to come up with a different workaround that is probably even uglier.

## One more alternative

We could also do the patch at link time with the `-ldflags -X` option, [similar to how we hard-code the version and commit](https://github.com/hashicorp/vault/blob/main/scripts/build.sh#L43) of Vault when building. The downside of this approach is that we can't easily write unit tests to verify that the patch worked.

## Testing

I checked out Vault 1.11 (`release/1.11.x`), upgraded to Go 1.19.3, and verified that the tests in `builtin/credential/aws/pkcs7/sign_test.go` that use SHA-1 failed.

I cherry-picked the commit in this PR, and verified that all the tests then passed.

## What next?

After this is merged, we can upgrade to the latest Go 1.19 in the 1.11.x and 1.10.x release branches in subsequent PRs.

### CODE REVIEW

File: command/commands.go [hunk 1]
```diff
@@ -206,6 +207,11 @@ var (
 	initCommandsEnt = func(ui, serverCmdUi cli.Ui, runOpts *RunOptions) {}
 )
 
+func init() {
```
Comment (swenson) [inline command/commands.go:210]: This *seemed* like the best place to put this, since it will always be run by the server when starting up, but I'm open to suggestion.

File: internal/go118_sha1_patch.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
+package internal
```
Comment (swenson) [inline internal/go118_sha1_patch.go:1]: This is put into a separate file so that it can be called by unit tests who can't import the `command` package.

File: internal/go118_sha1_patch.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+// PatchSha1 patches Go 1.18+ to allow certificates with signatures containing SHA-1 hashes to be allowed.
+// It is safe to call this function multiple times.
+// This is necessary to allow Vault 1.10 and 1.11 to work with Go 1.18+ without breaking backwards compatibility
+// with these certificates. See https://go.dev/doc/go1.18#sha1 and
+// https://developer.hashicorp.com/vault/docs/deprecation/faq#q-what-is-the-impact-of-removing-support-for-x-509-certificates-with-signatures-that-use-sha-1
+// for more details.
+// TODO: remove when Vault <=1.11 is no longer supported
+func PatchSha1() {
+	patchSha1.Do(func() {
```
Comment (swenson) [inline internal/go118_sha1_patch.go:28]: It seemed safer to only apply the patch once and only for Vault <=1.11.

File: internal/go118_sha1_patch.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+		patchBefore, err := goversion.NewSemver(sha1PatchVersionsBefore)
+		if err != nil {
+			panic(err)
```
Comment (tomcf-hcp) [inline internal/go118_sha1_patch.go:31]: Are we sure we want a panic here? Later you set `patch = true` if `err` is not `nil`.

File: internal/go118_sha1_patch.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+		}
+
+		patch := false
+		v, err := goversion.NewSemver(version.GetVersion().Version)
+		if err == nil {
+			patch = v.LessThan(patchBefore)
+		} else {
+			fmt.Fprintf(os.Stderr, "Cannot parse version %s; going to apply SHA-1 deprecation patch workaround\n", version.GetVersion().Version)
+			patch = true
```
Comment (tomcf-hcp) [inline internal/go118_sha1_patch.go:40]: Should the default be false, rather than true?

File: internal/go118_sha1_patch.go (hunk 4 above)
Comment (swenson) [reply internal/go118_sha1_patch.go:31]: `sha1PatchVersionsBefore` is a `const`, so this should always succeed. If it doesn't, I think we'd rather know about it during unit tests than have it fail.

File: internal/go118_sha1_patch.go (hunk 5 above)
Comment (swenson) [reply internal/go118_sha1_patch.go:40]: I could see an argument either way honestly.

File: internal/go118_sha1_patch.go (hunk 4 above)
Comment (tomcf-hcp) [reply internal/go118_sha1_patch.go:31]: Fair. But then why not panic in the `else` on line 38?

Comment (swenson) [reply internal/go118_sha1_patch.go:31]: Mostly because that value isn't a `const`, since it isn't set until we build and link the binary. I could panic there too. I just slightly prefer to fail a little more safely if the user sets the version incorrectly.

File: command/commands.go (hunk 1 above)
Comment (calvn) [reply command/commands.go:210]: You could invoke this in `core.postUnseal`, but perhaps @ncabatoff has a better suggestion.

Comment (swenson) [reply command/commands.go:210]: I was trying to target somewhere that will also be run when Agent starts up. (Ideally, also when any plugin starts up in a separate process, but that would require hooking into `sdk/` I think.)

**Comment by mladlow** [comment]
I wound up getting linked to this PR via some other means and just wanted to say it's a great writeup @swenson - thanks for that!

File: command/commands.go (hunk 1 above)
Comment (tomhjp) [reply command/commands.go:210]: This seems perfectly fine to me. The other obvious place is `main.go` in the root of the repo, but maybe I'm missing a reason that wouldn't work.

I think you could also just put the `init` function directly in the internal module and include it as an unnamed import (`_ "github.com/hashicorp/vault/internal"`). That would net a couple of benefits:
* Eliminates all exported API surface to prevent other calls
* Go's `init` function semantics will handle ensuring the function is only ever called once, and before everything else runs, so e.g. you can do away with the `sync.Once` usage.

File: internal/go118_sha1_patch.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+// for more details.
+// TODO: remove when Vault <=1.11 is no longer supported
+func PatchSha1() {
+	patchSha1.Do(func() {
+		patchBefore, err := goversion.NewSemver(sha1PatchVersionsBefore)
+		if err != nil {
+			panic(err)
+		}
+
+		patch := false
```
Comment (tomhjp) [inline internal/go118_sha1_patch.go:34]: I would prefer to PR this change directly into the 1.11.x and 1.10.x release branches, and then we don't need any logic about whether to patch.

Comment (swenson) [reply internal/go118_sha1_patch.go:34]: re: putting it directly in 1.11.x vs. 1.10.x, I slightly prefer this way, to keep future backports cleaner. But, changing directly in the 1.11.x and 1.10.x branches is okay, it's just slightly more work, and could create more conflicts.

re: where to put it, the `main.go` makes sense (and I'll move the init there). But, I wanted to keep it available for plugins though, i.e., if someone were to build a built-in plugin to run externally, they should be able to run the patch as well, which would mean it can't *only* be available in the root `main.go`.

**Review by tvoran — APPROVED** [review]

**Comment by swenson** [comment]
Thanks everyone!

**Review by benashz — COMMENTED** [review]
LGTM!
